### PR TITLE
Handle mkldnn broadcasting in mkldnn fuser

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* #51921 Add 0-dim support for binary mkldnn ops
* **#51736 Handle mkldnn broadcasting in mkldnn fuser**
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Differential Revision: [D26696694](https://our.internmc.facebook.com/intern/diff/D26696694)

### CODE REVIEW

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 1]
```diff
@@ -300,9 +355,9 @@ class MKLDNNSubgraphSlicer {
       return supportedMKLDNNWeight(*const_tensor);
     }
     auto k = v->node()->kind();
-    bool supported_kind = k == prim::MKLDNNGroup || k == aten::to_mkldnn ||
-        k == prim::ConstantMKLDNNTensor;
-    if (supported_kind) {
+    bool supported = k == prim::MKLDNNGroup ||
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:358]: nitpick: if that "if" statement is the only use of `supported` maybe we should just inline it?

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp [hunk 2]
```diff
@@ -34,6 +34,41 @@ using Tensor = at::Tensor;
...
+    auto b_size = b.sizes();
+    auto a_size = a.sizes();
+    if (a_size.equals(b_size)) {
+      push(stack, a, b);
+    } else {
+      auto out_size = at::infer_size(a_size, b_size);
+      if (a_size.equals(out_size)) {
+        push(stack, a);
+      } else {
+        push(stack, a.to_dense().expand(out_size).to_mkldnn());
```
Comment (Krovatkin) [inline torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:54]: does this do physical broadcasting?

File: test/jit/test_freezing.py [hunk 3]
```diff
@@ -1627,6 +1627,38 @@ def test_unsupported(module, preserved_attrs=None):
...
+            scripted_mod = torch.jit.freeze(scripted_mod)
+            self.run_pass("convert_frozen_ops_to_mkldnn", scripted_mod.graph)
+            FileCheck().check("prim::BroadcastMKLDNNTensors").run(scripted_mod.graph)
+            inp = torch.rand([20, 20])
+            self.assertEqual(scripted_mod(inp), mod(inp))
+            self.assertEqual(scripted_mod(inp), mod(inp))
+
+            # for good measure, check that broadcasting does not work without this op
+            # so we can remove the op if it ever gets supported
+            with self.assertRaisesRegex(RuntimeError, ""):
```
Comment (Krovatkin) [inline test/jit/test_freezing.py:1658]: 👍

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/frozen_ops_to_mkldnn.cpp:54]: yes

**Pull request closed without merging**